server/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java

```diff
@@ -204,8 +214,24 @@ private void releaseDelayedOperations() {
      * @param onAcquired      {@link ActionListener} that is invoked once acquisition is successful or failed
      * @param executorOnDelay executor to use for the possibly delayed {@link ActionListener#onResponse(Object)} call
      * @param forceExecution  whether the runnable should force its execution in case it gets rejected
+     * @param debugInfo       an extra information that can be useful when tracing an unreleased permit. When assertions are enabled
+     *                        the tracing will capture the supplied object's {@link Object#toString()} value. Otherwise the object
+     *                        isn't used
+     *
      */
-    public void acquire(final ActionListener<Releasable> onAcquired, final String executorOnDelay, final boolean forceExecution) {
+    public void acquire(final ActionListener<Releasable> onAcquired, final String executorOnDelay, final boolean forceExecution,
+                        final Object debugInfo) {
+        final Throwable debugInfoWithStackTrace;
+        if (Assertions.ENABLED) {
+            debugInfoWithStackTrace = new Throwable(debugInfo.toString());
```

Comment:
Is it possible to do this without a `Throwable`? Can you use a `Tuple` of `Object, StackTraceElement[]` populated with `Thread.currentThread().getStackTrace()` (and as an added bonanza avoid rendering the `Object` to a `String` unless absolutely needed)?